gdnative-core/src/nativescript/class.rs

```diff
@@ -155,117 +169,228 @@ impl<T: NativeClass> Instance<T> {
-    /// that it will not be freed until this function returns. Otherwise, it is undefined
-    /// behavior to call this function and/or use its return value.
+    /// It's safe to call `assume_safe` if and only if the constraints of `Ptr::assume_safe`
+    /// are satisfied for the base object.
     #[inline]
-    pub unsafe fn try_from_unsafe_base(owner: T::Base) -> Option<Self> {
-        let type_tag = (get_api().godot_nativescript_get_type_tag)(owner.to_sys());
-        if type_tag.is_null() {
-            return None;
+    pub unsafe fn assume_safe<'a>(self) -> RefInstance<'a, T> {
+        RefInstance {
+            owner: self.owner.assume_safe::<'a>(),
+            script: self.script,
         }
+    }
+
+    /// Assume that `self` to use during the lifetime of the input reference. The input
+    /// reference is unused and may be anything. This is a convenience wrapper around
+    /// `assume_safe`.
+    ///
+    /// This is *not* guaranteed to be a no-op at runtime.
+    ///
+    /// # Safety
+    ///
+    /// It's safe to call `assume_safe_during` if and only if the constraints of
+    /// `Ptr::assume_safe` are satisfied for the base object.
+    #[inline]
+    pub unsafe fn assume_safe_during<'a, L: 'a>(self, _lifetime: &'a L) -> RefInstance<'a, T> {
+        self.assume_safe::<'a>()
+    }
+
+    /// Manually frees the base object of this instance.
+    ///
+    /// # Safety
+    ///
+    /// During the call, the underlying object must be valid, and this thread must have
+    /// exclusive access to the object.
+    #[inline]
+    pub unsafe fn free(self) {
+        self.owner.free();
+    }
+}
+
+impl<'a, T: NativeClass> RefInstance<'a, T> {
+    /// Returns a reference to the base object with the same lifetime.
+    #[inline]
+    pub fn base(&self) -> &'a T::Base {
+        self.owner
+    }
+
+    /// Returns a reference to the script wrapper.
+    #[inline]
+    pub fn script(&self) -> &T::UserData {
+        &self.script
+    }
+
+    /// Try to downcast `&T::Base` to `RefInstance<T>`.
+    #[inline]
+    pub fn try_from_base(owner: &'a T::Base) -> Option<Self> {
+        unsafe {
+            let api = get_api();
+
+            let owner_ptr = owner.as_ptr();
+
+            let type_tag = (api.godot_nativescript_get_type_tag)(owner_ptr);
+            if type_tag.is_null() {
+                return None;
+            }
 
-        if !crate::nativescript::type_tag::check::<T>(type_tag) {
-            return None;
+            if !crate::nativescript::type_tag::check::<T>(type_tag) {
+                return None;
+            }
+
+            let user_data = (api.godot_nativescript_get_userdata)(owner_ptr);
+            Some(Self::from_raw_unchecked(owner, user_data))
         }
+    }
+
+    /// Pairs an `owner` and `user_data` without checking validity. Internal interface.
+    #[doc(hidden)]
+    #[inline]
+    pub unsafe fn from_raw_unchecked(owner: &'a T::Base, user_data: *mut libc::c_void) -> Self {
```

Comment:
I think this is used in the method macro currently. Might not be actually necessary, though. I'll refactor a bit more once the interface settles down.

Reply:
This is used in the method macro because the owner pointer and `user_data` pointer are passed separately for methods, and we assume that it's a valid pairing.